### ISSUE TITLE
snapshot atomicity

### ISSUE THREAD

**Issue by silenius** [zrepl/zrepl#634]
Currently zrepl makes independent snapshots for each child and grand-child. This is very flexible but does not scale to thousands of child-datasets and could even be problematic in some cases. A case that comes to mind is PostgreSQL with separate datasets for data and WAL (within the same pool), independant snapshots for WAL and data could result in corruption.

ZFS supports two options for recursive snapshots:
```bash
$> zfs snapshot -r zroot/mydataset@now
$> zfs snapshot zroot/pg13/data@now zroot/pg13/wal@now zroot/myotherdataset@now
```

I think it would be interresting if zrepl could support that

**Comment by notr1ch** [comment]
I thought this was already the case according to https://zrepl.github.io/configuration/snapshotting.html#periodic-snapshotting

> It is desirable to get all filesystems snapshotted simultaneously because it results in a more consistent backup. To accomplish this while still maintaining the interval, the periodic snapshotter attempts to get the snapshotting rhythms in sync.

**Comment by InsanePrawn** [comment]
> I thought this was already the case according to https://zrepl.github.io/configuration/snapshotting.html#periodic-snapshotting
> 
> > It is desirable to get all filesystems snapshotted simultaneously because it results in a more consistent backup. To accomplish this while still maintaining the interval, the periodic snapshotter attempts to get the snapshotting rhythms in sync.

That's talking about zrepl trying to synchronise timers, not using `zfs snap -r`.

As answered [here](https://github.com/zrepl/zrepl/discussions/632#discussioncomment-3726490), zrepl doesn't use `-r` and can't do so anyway, once you have even a single dataset excluded.

If you try to piece together a cmdline like `$> zfs snapshot zroot/pg13/data@now zroot/pg13/wal@now zroot/myotherdataset@now`, you run into problems with argv length when you have [long] enough datasets [names].

On top of that, it's unclear how hooks would interact with that.

Can you please explain why you need this feature?
Is a backup that's a few seconds apart suddenly less valuable?
Since I can't come up with the use case  for it, it sounds like a "i want it cause it looks good on paper" type of thing to me. 
Happy to be proven wrong though :)

**Comment by notr1ch** [comment]
Snapshots of an active database as mentioned in the OP is a good use case. I also have such a setup - separate datasets for the data files and log due to the different write characteristics. If those aren't snapshot atomically, data loss on a restore could occur due to the inconsistencies - the log may think it has written data that ended up not being in the snapshot for example.

This is the OpenZFS recommended way of running both MySQL and Postgres: https://openzfs.github.io/openzfs-docs/Performance%20and%20Tuning/Workload%20Tuning.html#database-workloads so it's not a niche use-case.

**Comment by InsanePrawn** [comment]
This databse use-case sounds like you'd want them to run their pre and post hook[s] collectively rather than -r your whole pool. Or would you rather have:
1. zrepl run *all* the hooks (i.e. all the DB pauses) for all the dataets on the pool
2. do one `zfs snapshot -r`
3. run all the post hooks on the pool

AFAIU what you really want is:
1. zrepl to run the pause hook for one database
2. snapshot all the datasets for that database, atomically or not
3. run the resume hook

I actually know very little about the SQL server side of things, so please correct me if I'm wrong.

**Comment by problame** [comment]
@notr1ch IIUC you are using two datasets, one for the Postgres WAL, one for the remaining Postgres on-disk state.

My remarks on this:

1. Mandatory reading: https://www.postgresql.org/docs/current/backup-file.html
2. Note that ZFS's filesystem-level commit log (ZIL) is per-dataset, so, in order for Postgres recovery to work after a crash, ZIL replay must succeed on _both_ datasets. Otherwise, esp if the main dataset's ZIL can't be replayed to the latest Postgres checkpoint, Postgres can't replay the WAL.
3. As stated in the doc linked above, if you snapshot with `zfs snap -r`, and then restore from those snapshots, Postgres will need to WAL-replay everything in the WAL that's after the last Postgres checkpoint. It increases recovery time. I suppose you're aware of that.

My question to you: would you be satisfied if zrepl offered atomic snapshots, but only if
1. there are no hooks for the job (would be a config error at startup), and
2. all `filesystems` of the job are in the same zpool, and
3. the user requested it via a config knob, like

```
- type: push
  ...
  filesystems: { "zroot<": true } 
  snapshotting:
    atomic:
      - { "zroot/pg/wal": true, "zroot/pg/data": true }
      - { "zroot/mariadb/wal": true, "zroot/mariadb/data": true }

```

**Comment by notr1ch** [comment]
Yes, personally I would be satisfied with such an implementation.

If it doesn't complicate things I think still being able to support hooks would be a useful feature - for example while MySQL+InnoDB is snapshot-safe, other MySQL table formats may not be, so a hook to issue FLUSH TABLES / BACKUP STAGE could still be useful in some situations. In the case of atomic snapshots, I would imagine the pre-hooks to fire on all the atomic-grouped filesystems (potentially in an undefined order), then the snapshot, then the post-hooks.

**Comment by silenius** [comment]
> > I thought this was already the case according to https://zrepl.github.io/configuration/snapshotting.html#periodic-snapshotting
> > > It is desirable to get all filesystems snapshotted simultaneously because it results in a more consistent backup. To accomplish this while still maintaining the interval, the periodic snapshotter attempts to get the snapshotting rhythms in sync.
> 
> That's talking about zrepl trying to synchronise timers, not using `zfs snap -r`.
> 
> As answered [here](https://github.com/zrepl/zrepl/discussions/632#discussioncomment-3726490), zrepl doesn't use `-r` and can't do so anyway, once you have even a single dataset excluded.
> 
> If you try to piece together a cmdline like `$> zfs snapshot zroot/pg13/data@now zroot/pg13/wal@now zroot/myotherdataset@now`, you run into problems with argv length when you have [long] enough datasets [names].
> 
> On top of that, it's unclear how hooks would interact with that.
> 
> Can you please explain why you need this feature? Is a backup that's up to a few seconds apart suddenly less valuable? Since I can't come up with the use case for it, it sounds like a "i want it cause it looks good on paper" type of thing to me. Happy to be proven wrong though :)

Apart from database servers, a use case is any application that stores it's data to a dedicated dataset. In my case I'm using FreeBSD jails widely and I always use a dedicated dataset to store data generated by the jailed application. This greatly facilitates the upgrade of the jails as I could zfs destroy /thejail/dataset and recreate it in a few seconds with some CMS (I use Saltstack) without having to care about the data.

**Comment by silenius** [comment]
> @notr1ch IIUC you are using two datasets, one for the Postgres WAL, one for the remaining Postgres on-disk state.
> 
> My remarks on this:
> 
>     1. Mandatory reading: https://www.postgresql.org/docs/current/backup-file.html
> 
>     2. Note that ZFS's filesystem-level commit log (ZIL) is per-dataset, so, in order for Postgres recovery to work after a crash, ZIL replay must succeed on _both_ datasets. Otherwise, esp if the main dataset's ZIL can't be replayed to the latest Postgres checkpoint, Postgres can't replay the WAL.
> 
>     3. As stated in the doc linked above, if you snapshot with `zfs snap -r`, and then restore from those snapshots, Postgres will need to WAL-replay everything in the WAL that's after the last Postgres checkpoint. It increases recovery time. I suppose you're aware of that.
> 
> 
> My question to you: would you be satisfied if zrepl offered atomic snapshots, but only if
> 
>     1. there are no hooks for the job (would be a config error at startup), and
> 
>     2. all `filesystems` of the job are in the same zpool, and
> 
>     3. the user requested it via a config knob, like
> 
> 
> ```
> - type: push
>   ...
>   filesystems: { "zroot<": true } 
>   snapshotting:
>     atomic:
>       - { "zroot/pg/wal": true, "zroot/pg/data": true }
>       - { "zroot/mariadb/wal": true, "zroot/mariadb/data": true }
> ```

I would also be satisfied with such an implementation :+1:

**Comment by lonjil** [comment]
Any plans to support simply passing multiple datasets to `zfs snapshot` instead of using `-r`? All snapshots made in a single invocation of the command go into the same txg, just like with `-r`.
